{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.metrics import roc_auc_score, r2_score, accuracy_score, f1_score, precision_score, recall_score, log_loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from explainerdashboard.explainers import *\n",
    "from explainerdashboard.dashboards import *\n",
    "from explainerdashboard.datasets import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RandomForestClassifier dashboard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, y_train, X_test, y_test = titanic_survive()\n",
    "train_names, test_names = titanic_names()\n",
    "\n",
    "model = RandomForestClassifier(n_estimators=50, max_depth=5)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "explainer = RandomForestClassifierBunch(model, X_test, y_test, roc_auc_score, \n",
    "                               cats=['Sex', 'Deck', 'Embarked'],\n",
    "                               idxs=test_names, #names of passengers \n",
    "                               labels=['Not survived', 'Survived'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "hoverinfo": "text",
         "insidetextanchor": "end",
         "orientation": "h",
         "text": [
          0,
          0,
          0,
          0,
          0.001,
          0.001,
          0.002,
          0.002,
          0.002,
          0.004,
          0.007,
          0.007,
          0.007,
          0.008,
          0.011,
          0.013,
          0.015,
          0.038,
          0.038,
          0.044,
          0.071,
          0.109
         ],
         "textposition": "inside",
         "type": "bar",
         "x": [
          0,
          2.691828833540237e-05,
          6.506241686417105e-05,
          0.0002421284605699875,
          0.0006707882017245051,
          0.001137569574206446,
          0.0015574918561918976,
          0.0018554581968534204,
          0.0024206138087656884,
          0.00424274994414861,
          0.006636417287191547,
          0.006915432679575293,
          0.007168143441198895,
          0.008445560033483814,
          0.011295045369277825,
          0.013015309916663545,
          0.015056312921864306,
          0.0377027567084739,
          0.03771918875205964,
          0.04381405172507327,
          0.07129681028066447,
          0.10912407599287235
         ],
         "y": [
          "22. Sex_nan",
          "21. Embarked_Unknown",
          "20. Deck_T",
          "19. Deck_G",
          "18. Deck_A",
          "17. Deck_C",
          "16. Deck_F",
          "15. Embarked_Queenstown",
          "14. Deck_D",
          "13. Deck_E",
          "12. No_of_siblings_plus_spouses_on_board",
          "11. Deck_B",
          "10. No_of_parents_plus_children_on_board",
          "9. Embarked_Southamption",
          "8. Embarked_Cherbourg",
          "7. Age",
          "6. No_of_relatives_on_board",
          "5. Deck_Unkown",
          "4. PassengerClass",
          "3. Fare",
          "2. Sex_female",
          "1. Sex_male"
         ]
        }
       ],
       "layout": {
        "height": 640,
        "margin": {
         "b": 50,
         "l": 252,
         "pad": 4,
         "r": 50,
         "t": 50
        },
        "plot_bgcolor": "#fff",
        "showlegend": false,
        "template": {
         "data": {
          "scatter": [
           {
            "type": "scatter"
           }
          ]
         }
        },
        "title": {
         "text": "MEAN_ABS_SHAP"
        },
        "xaxis": {
         "automargin": true
        },
        "yaxis": {
         "automargin": true
        }
       }
      },
      "text/html": [
       "<div>\n",
       "        \n",
       "        \n",
       "            <div id=\"4d8c1520-fced-4e19-a5e3-0026880c540d\" class=\"plotly-graph-div\" style=\"height:640px; width:100%;\"></div>\n",
       "            <script type=\"text/javascript\">\n",
       "                require([\"plotly\"], function(Plotly) {\n",
       "                    window.PLOTLYENV=window.PLOTLYENV || {};\n",
       "                    \n",
       "                if (document.getElementById(\"4d8c1520-fced-4e19-a5e3-0026880c540d\")) {\n",
       "                    Plotly.newPlot(\n",
       "                        '4d8c1520-fced-4e19-a5e3-0026880c540d',\n",
       "                        [{\"hoverinfo\": \"text\", \"insidetextanchor\": \"end\", \"orientation\": \"h\", \"text\": [0.0, 0.0, 0.0, 0.0, 0.001, 0.001, 0.002, 0.002, 0.002, 0.004, 0.007, 0.007, 0.007, 0.008, 0.011, 0.013, 0.015, 0.038, 0.038, 0.044, 0.071, 0.109], \"textposition\": \"inside\", \"type\": \"bar\", \"x\": [0.0, 2.691828833540237e-05, 6.506241686417105e-05, 0.0002421284605699875, 0.0006707882017245051, 0.001137569574206446, 0.0015574918561918976, 0.0018554581968534204, 0.0024206138087656884, 0.00424274994414861, 0.006636417287191547, 0.006915432679575293, 0.007168143441198895, 0.008445560033483814, 0.011295045369277825, 0.013015309916663545, 0.015056312921864306, 0.0377027567084739, 0.03771918875205964, 0.04381405172507327, 0.07129681028066447, 0.10912407599287235], \"y\": [\"22. Sex_nan\", \"21. Embarked_Unknown\", \"20. Deck_T\", \"19. Deck_G\", \"18. Deck_A\", \"17. Deck_C\", \"16. Deck_F\", \"15. Embarked_Queenstown\", \"14. Deck_D\", \"13. Deck_E\", \"12. No_of_siblings_plus_spouses_on_board\", \"11. Deck_B\", \"10. No_of_parents_plus_children_on_board\", \"9. Embarked_Southamption\", \"8. Embarked_Cherbourg\", \"7. Age\", \"6. No_of_relatives_on_board\", \"5. Deck_Unkown\", \"4. PassengerClass\", \"3. Fare\", \"2. Sex_female\", \"1. Sex_male\"]}],\n",
       "                        {\"height\": 640, \"margin\": {\"b\": 50, \"l\": 252, \"pad\": 4, \"r\": 50, \"t\": 50}, \"plot_bgcolor\": \"#fff\", \"showlegend\": false, \"template\": {\"data\": {\"scatter\": [{\"type\": \"scatter\"}]}}, \"title\": {\"text\": \"MEAN_ABS_SHAP\"}, \"xaxis\": {\"automargin\": true}, \"yaxis\": {\"automargin\": true}},\n",
       "                        {\"responsive\": true}\n",
       "                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('4d8c1520-fced-4e19-a5e3-0026880c540d');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })\n",
       "                };\n",
       "                });\n",
       "            </script>\n",
       "        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "explainer.plot_importances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating shap values...\n",
      "Generating shap TreeExplainer...\n",
      "Calculating predictions...\n",
      "Calculating ranks...\n",
      "Calculating prediction probabilities...\n",
      "Calculating importances...\n",
      "Generating ShadowDecTree for each individual decision tree...\n",
      "Running Model Explainer on http://localhost:8052\n",
      " * Serving Flask app \"explainerdashboard.dashboards\" (lazy loading)\n",
      " * Environment: production\n",
      "   WARNING: This is a development server. Do not use it in a production deployment.\n",
      "   Use a production WSGI server instead.\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:8052/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:45] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:45] \"GET /_dash-dependencies HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:45] \"GET /_dash-layout HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:46] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:46] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:46] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:46] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:46] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 204 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 204 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 204 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\model_summary_tab.py\", line 131, in update_importances\n",
      "    kind=permutation_shap, topx=tablesize, cats=cats)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 684, in plot_importances\n",
      "    return plotly_importances_plot(importances_df, round=round)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainer_plots.py\", line 808, in plotly_importances_plot\n",
      "    importance_name = importance_df.columns[1] # can be \"MEAN_ABS_SHAP\", \"Permutation Importance\", etc\n",
      "AttributeError: 'NoneType' object has no attribute 'columns'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:49] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:50] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:50] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:51] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:52] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1457, in dispatch\n",
      "    response.set_data(self.callback_map[output][\"callback\"](*args))\n",
      "  File \"C:\\ProgramData\\Anaconda3\\envs\\ww_env2\\lib\\site-packages\\dash\\dash.py\", line 1337, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\dashboard_tabs\\contributions_tab.py\", line 283, in update_pdp_graph\n",
      "    return explainer.plot_pdp(col, idx, sample=100)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 833, in plot_pdp\n",
      "    num_grid_points=num_grid_points)\n",
      "  File \"C:\\Users\\odi011\\projects\\explainerdashboard\\explainerdashboard\\explainers.py\", line 1233, in get_pdp_result\n",
      "    pdp_result = super(ClassifierBunch, self).get_pdp_result(\n",
      "TypeError: super(type, obj): obj must be an instance or subtype of type\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [28/Feb/2020 11:24:52] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:52] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:52] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:52] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:53] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:54] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:54] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [28/Feb/2020 11:24:54] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    }
   ],
   "source": [
    "db = ExplainerDashboard(explainer,\n",
    "                        model_summary=True,\n",
    "                        contributions=True,\n",
    "                        shap_dependence=False,\n",
    "                        shap_interaction=False,\n",
    "                        decision_trees=True)\n",
    "db.run(8052)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RandomForestRegressor example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, y_train, X_test, y_test = titanic_fare()\n",
    "train_names, test_names = titanic_names()\n",
    "\n",
    "model = RandomForestRegressor(n_estimators=50, max_depth=5)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "explainer = RandomForestRegressionBunch(model, X_test, y_test, r2_score,\n",
    "                               cats=['Sex', 'Deck', 'Embarked'],\n",
    "                               idxs=test_names, units=\"$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = ExplainerDashboard(explainer,\n",
    "                        model_summary=True,\n",
    "                        contributions=True,\n",
    "                        shap_dependence=True,\n",
    "                        shap_interaction=True,\n",
    "                        shadow_trees=True)\n",
    "db.run(8052)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XGBClassifier example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, y_train, X_test, y_test = titanic_survive()\n",
    "train_names, test_names = titanic_names()\n",
    "\n",
    "xgb_model = XGBClassifier()\n",
    "xgb_model.fit(X_train, y_train)\n",
    "\n",
    "explainer = TreeClassifierBunch(xgb_model, X_test, y_test, roc_auc_score,\n",
    "                                   cats=['Sex', 'Cabin', 'Embarked'],\n",
    "                                   idxs=test_names, #names of passengers \n",
    "                                   labels=['Not survived', 'Survived'])\n",
    "\n",
    "db = ExplainerDashboard(explainer,\n",
    "                        model_summary=True,\n",
    "                        contributions=True,\n",
    "                        shap_dependence=True,\n",
    "                        shap_interaction=False)\n",
    "db.run(8052)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
